File: ffi/poc/rust/crates/harness/src/bin/counts.rs
//! Boundary-call counts per payload per direction, from a counting build (README R5).
//!
//! Counted inside the core, at the entry points and at each function-pointer invocation, so
//! what is reported is what happened rather than what the shape of the code suggests. R5
//! also requires the boundary itself be shown to exist, which `gen/stage3.sh` does with
//! `nm -D --undefined-only`: a counting build alone is not evidence that a call happened,
//! because an inlined body carries its counters with it.
//!
//! The `class` column is per design/SHAPES.md: a message the real schema barely has is a
//! CONTROL and its row is not a verdict. It is labelled here and not only in SHAPES.md,
//! because reports get assembled from logs.

use ak_abi::*;
use harness::arms_m2;
use harness::arms::*;

fn main() {
    println!("# boundary-call counts, arm core-ffi-rust");
    println!("#   counting build: {}", if cfg!(feature = "count") { "YES" } else { "no (figures below are meaningless)" });
    println!("#   guard:          {}", if cfg!(feature = "guard") { "on" } else { "OFF" });
    println!();
    println!(
        "{:<7} {:<8} {:<9} {:>9} {:>8} {:>8} {:>10} {:>7} {:>16} {:>6}",
        "payload", "class", "direction", "elements", "forward", "reverse", "crossings", "transc", "prefix", "grows"
    );
    let mut regressions = 0usize;

    let c = core_ffi_arm::Ctx::new();

    // ---- M1, a real message: 3 scalars, 6 strings, 2 singular nested messages.
    for pid in [P1_1, P1_2, P1_3] {
        let v = core_ffi_arm::value(pid);
        let n = v.results.len();
        unsafe {
            let cold = core_ffi_arm::Ctx::new();
            core_ffi_arm::encode_into(&cold, &v);
            row(pid, "real", "enc cold", n, &take_enc(cold.enc));

            core_ffi_arm::encode_into(&c, &v);
            ak_enc_counters_reset(c.enc);
            core_ffi_arm::encode_into(&c, &v);
            row(pid, "real", "enc warm", n, &take_enc(c.enc));

            let b = core_ffi_arm::encode(&c, &v);
            core_ffi_arm::decode(&c, &b);
            ak_dec_counters_reset(c.dec);
            core_ffi_arm::decode(&c, &b);
            row(pid, "real", "decode", n, &take_dec(c.dec));
        }
    }

    // ---- M2, a real message, and the one the control plane actually moves.
    for pid in arms_m2::ALL {
        let v = arms_m2::core_ffi_arm::value(pid);
        let n = v.tasks.len();
        unsafe {
            let cold = core_ffi_arm::Ctx::new();
            arms_m2::core_ffi_arm::encode_into(&cold, &v);
            row(pid, "real", "enc cold", n, &take_enc(cold.enc));

            arms_m2::core_ffi_arm::encode_into(&c, &v);
            ak_enc_counters_reset(c.enc);
            arms_m2::core_ffi_arm::encode_into(&c, &v);
            row(pid, "real", "enc warm", n, &take_enc(c.enc));

            let b = arms_m2::core_ffi_arm::encode(&c, &v);
            arms_m2::core_ffi_arm::decode(&c, &b);
            ak_dec_counters_reset(c.dec);
            arms_m2::core_ffi_arm::decode(&c, &b);
            row(pid, "real", "decode", n, &take_dec(c.dec));
        }
    }

    // ---- M3: a oneof and explicit presence. Both ride in the group, so neither adds a
    // crossing; that is the point and the count is how it is shown rather than asserted.
    {
        use harness::arms_m3 as m3;
        let v = m3::core_ffi_arm::value(m3::P3_1);
        let n = v.probes.len();
        unsafe {
            let cold = core_ffi_arm::Ctx::new();
            m3::core_ffi_arm::encode_into(&cold, &v);
            row("P3.1", "real", "enc cold", n, &take_enc(cold.enc));
            m3::core_ffi_arm::encode_into(&c, &v);
            ak_enc_counters_reset(c.enc);
            m3::core_ffi_arm::encode_into(&c, &v);
            row("P3.1", "real", "enc warm", n, &take_enc(c.enc));
            let b = m3::core_ffi_arm::encode(&c, &v);
            m3::core_ffi_arm::decode(&c, &b);
            ak_dec_counters_reset(c.dec);
            m3::core_ffi_arm::decode(&c, &b);
            row("P3.1", "real", "decode", n, &take_dec(c.dec));
        }
    }

    // ---- ABI v1 open decision 5, by site. An aggregate that does not say WHERE the
    // learned width thrashes cannot answer what the grow path costs.
    println!();
    println!("# length-prefix misses by site, warm context (ABI v1 open decision 5)");
    println!("#   a site is one length-prefix call site in the generated codec. Its learned");
    println!("#   width is per context and starts at one byte.");
    println!();
    for pid in ["P1.2", "P2.2", "P2.4"] {
        let ctx = core_ffi_arm::Ctx::new();
        unsafe {
            if pid == "P1.2" {
                let v = core_ffi_arm::value(pid);
                core_ffi_arm::encode_into(&ctx, &v);
                ak_enc_counters_reset(ctx.enc);
                let mut z = vec![0u32; SITE_NAMES.len()];
                ak_enc_site_moves(ctx.enc, z.as_mut_ptr(), z.len());
                let before = z.clone();
                core_ffi_arm::encode_into(&ctx, &v);
                ak_enc_site_moves(ctx.enc, z.as_mut_ptr(), z.len());
                site_report(pid, v.results.len(), &before, &z);
            } else {
                let v = arms_m2::core_ffi_arm::value(pid);
                arms_m2::core_ffi_arm::encode_into(&ctx, &v);
                let mut z = vec![0u32; SITE_NAMES.len()];
                ak_enc_site_moves(ctx.enc, z.as_mut_ptr(), z.len());
                let before = z.clone();
                arms_m2::core_ffi_arm::encode_into(&ctx, &v);
                ak_enc_site_moves(ctx.enc, z.as_mut_ptr(), z.len());
                site_report(pid, v.tasks.len(), &before, &z);
            }
        }
    }

    // A regression that would have caught the open-state leak: on a payload whose elements
    // are all the same shape, a learned width converges after the first pass and a warm
    // context must never miss again. It missed 448 times in 500 elements because the second
    // and later chunks of an element run read whatever the previous element left in the
    // codec's open-field state.
    println!();
    println!("# regression: a warm context must not miss on a UNIFORM payload");
    for pid in ["P1.2", "P2.2", "P2.3", "P2.5"] {
        let ctx = core_ffi_arm::Ctx::new();
        let misses = unsafe {
            if pid == "P1.2" {
                let v = core_ffi_arm::value(pid);
                core_ffi_arm::encode_into(&ctx, &v);
                ak_enc_counters_reset(ctx.enc);
                core_ffi_arm::encode_into(&ctx, &v);
            } else {
                let v = arms_m2::core_ffi_arm::value(pid);
                arms_m2::core_ffi_arm::encode_into(&ctx, &v);
                ak_enc_counters_reset(ctx.enc);
                arms_m2::core_ffi_arm::encode_into(&ctx, &v);
            }
            take_enc(ctx.enc).prefix_moves
        };
        println!("#   {pid:<6} warm misses = {misses} {}", if misses == 0 { "ok" } else { "REGRESSION" });
        if misses != 0 {
            regressions += 1;
        }
    }
    if regressions > 0 {
        println!();
        println!("FAILED: {regressions} uniform payload(s) do not converge.");
        std::process::exit(1);
    }

    println!();
    println!("# `crossings` is forward + reverse: calls that leave one side's compilation unit.");
    println!("# `transc` counts transcoder invocations, which are indirect calls the core makes");
    println!("#   into ITSELF (ABI v1 section 4 puts every representation in the core), so they");
    println!("#   are not host crossings. Counting them together is what made the drafted");
    println!("#   interface look like 15,137 calls on a 1,000-row response.");
    println!("# `prefix` counts length-placeholder resizes: ABI v1 open decision 5. `enc cold`");
    println!("#   is a context whose learned widths have never been set, `enc warm` the steady");
    println!("#   state. P2.4 alternates element bodies across a varint length boundary by");
    println!("#   construction, so a per-site learned width is wrong on every element of it:");
    println!("#   that row is the one decision 5 turns on.");
    println!("# M1's element type is a LEAF, so its rows batch. M2's is not (ABI v1 7.2), so");
    println!("#   M2 costs two calls per element plus one run per repeated field it carries.");
}

fn site_report(pid: &str, elems: usize, before: &[u32], after: &[u32]) {
    let mut rows: Vec<(u32, &str)> = SITE_NAMES
        .iter()
        .enumerate()
        .map(|(i, n)| (after[i].saturating_sub(before[i]), *n))
        .filter(|(d, _)| *d > 0)
        .collect();
    rows.sort_by(|a, b| b.0.cmp(&a.0));
    let total: u32 = rows.iter().map(|(d, _)| d).sum();
    if rows.is_empty() {
        println!("  {pid:<6} {elems} elements: no misses in the warm pass");
        return;
    }
    println!("  {pid:<6} {elems} elements, {total} misses in one warm pass:");
    for (d, n) in rows {
        println!("      {d:>6}  {:>7}  {n}", format!("{:.2}/el", d as f64 / elems as f64));
    }
}

unsafe fn take_enc(ctx: *mut ak_enc_ctx) -> AkCounters {
    let mut z = AkCounters::default();
    ak_enc_counters(ctx, &mut z);
    z
}

unsafe fn take_dec(ctx: *mut ak_dec_ctx) -> AkCounters {
    let mut z = AkCounters::default();
    ak_dec_counters(ctx, &mut z);
    z
}

fn row(pid: &str, class: &str, dir: &str, elems: usize, c: &AkCounters) {
    let cr = c.forward + c.reverse;
    println!(
        "{:<7} {:<8} {:<9} {:>9} {:>8} {:>8} {:>10} {:>7} {:>16} {:>6}",
        pid, class, dir, elems, c.forward, c.reverse, cr, c.transcode,
        if c.prefix_moves == 0 { "0".to_string() } else { format!("{} ({} B)", c.prefix_moves, c.prefix_bytes) },
        c.grows
    );
    if elems > 0 && dir != "enc cold" {
        println!(
            "{:<7} {:<8} {:<9} {:>9} {:>8} {:>8} {:>10} {:>7} {:>16} {:>6}",
            "", "", "  per elem", "", "", "", format!("{:.3}", cr as f64 / elems as f64), "", "", ""
        );
    }
}

File: ffi/poc/csharp/src/Harness/Counts.cs
// R5, and ABI v1 open decision 5.
//
// R5 asks for boundary-call counts per payload per direction, from a counting
// build. **Every arm in this slice is in-process managed code and crosses
// nothing**, so the honest count is zero, and this file says so rather than
// leaving the column blank. The counts R5 exists for belong to the `core-ffi`
// arm, which is not built because ABI v1 open decision 1 is unsettled.
//
// What IS counted here is the thing the managed codec can be counted for: the
// learned length-placeholder width of ABI v1 section 6, and how often it
// misses. Decision 5 asks what the mechanism is worth, and an aggregate that
// cannot say WHICH site thrashes does not answer it -- so the miss is recorded
// per site, and `Codec.SiteNames` names the field.
//
// The Rust slice's answer, for comparison and not for arithmetic: zero warm
// misses on every uniform payload; on P2.4, one miss per element, moving
// 980,938 of 981,222 bytes; about 1 to 3 percentage points of an encode on the
// payload built to defeat it, and nothing on any uniform one. P2.4 is the only
// payload whose element bodies alternate across a varint length boundary.

using System;
using System.Collections.Generic;
using System.Linq;
using Armonik.Ffi.Facade;

namespace Armonik.Ffi.Harness;

public static class Counts
{
    public static int Run(params string[] argv)
    {
        var rows = Manifest.Load();

        Console.WriteLine("Boundary crossings per payload per direction, every arm in this slice:");
        Console.WriteLine();
        Console.WriteLine("    gp-tobytearray   0     managed         0");
        Console.WriteLine("    gp-writeto       0     managed-2pass   0");
        Console.WriteLine("    gp-parse         0     managed-parse   0");
        Console.WriteLine();
        Console.WriteLine("  Not an estimate and not an omission: there is no boundary in any of them.");
        Console.WriteLine("  The core-ffi arm, which is the one R5's counts are about, is not built.");
        Console.WriteLine();
        Console.WriteLine("ABI v1 open decision 5: the learned length-placeholder width, and its misses.");
        Console.WriteLine();
        Console.WriteLine("payload  elements     bytes   cold misses   warm misses   warm bytes moved   grows");
        Console.WriteLine(new string('-', 98));

        var perSite = new Dictionary<string, long>();

        foreach (var a in ArmTable.All())
        {
            var row = rows[a.Id];
            a.Build();
            int cap = Math.Max(row.Bytes + 4096, 8192);
            var e = Enc.New(Codec.Sites, cap);

            // Cold: the widths all start at 1, so the first encode pays for
            // every site whose body is longer than 127 bytes.
            a.ManagedWrite(ref e);
            long cold = e.PrefixMoves;

            // Warm: the widths now carry what the last encode learned. This is
            // the steady state a server is in, and it is the number decision 5
            // is about.
            e.Reset();
            a.ManagedWrite(ref e);
            long warm = e.PrefixMoves, warmBytes = e.PrefixBytes, grows = e.Grows;

            Console.WriteLine("{0,-8} {1,8}  {2,8}  {3,12}  {4,12}  {5,17}  {6,6}",
                a.Id, a.Elements, row.Bytes, cold, warm, warmBytes, grows);

            if (warm > 0)
            {
                // Which site. Re-run with a per-site tally by encoding once per
                // site-isolating pass would cost 72 encodes; instead the site
                // names are printed for the payloads that miss, from a build
                // that tallies them directly.
                perSite[a.Id] = warm;
            }
        }

        Console.WriteLine();
        Console.WriteLine("  cold misses:  the first encode on a fresh context, every width still 1.");
        Console.WriteLine("  warm misses:  the second encode, widths carrying what the first learned. A");
        Console.WriteLine("                server is in this state; the cold column is a startup cost paid");
        Console.WriteLine("                once per context.");
        Console.WriteLine("  grows:        buffer reallocations. The harness gives the context a buffer big");
        Console.WriteLine("                enough for the payload, so a nonzero grow here is a defect in the");
        Console.WriteLine("                sizing and not a property of the codec.");
        Console.WriteLine();
        if (perSite.Count == 0)
        {
            Console.WriteLine("  No payload misses warm. Including P2.4, which is the payload BUILT to defeat");
            Console.WriteLine("  a learned width: its element bodies alternate 3 and 150 repeated strings, so");
            Console.WriteLine("  a per-call-site learned width is wrong on every element of it by construction.");
            Console.WriteLine("  If that row reads zero, read it as a defect in this harness before reading it");
            Console.WriteLine("  as a result -- see the per-site breakdown below.");
        }
        else
        {
            Console.WriteLine("  Payloads that miss warm: {0}", string.Join(", ",
                perSite.Select(kv => kv.Key + " (" + kv.Value + ")")));
        }

        Console.WriteLine();
        Console.WriteLine("Per-site warm misses on P2.4, the payload built to defeat a learned width.");
        Console.WriteLine();
        PerSite("P2.4", rows);
        Console.WriteLine();
        Console.WriteLine("Per-site warm misses on P2.2, the shape the control plane actually moves.");
        Console.WriteLine();
        PerSite("P2.2", rows);
        return 0;
    }

    /// The per-site tally, taken by encoding twice and diffing the width table:
    /// a site whose learned width CHANGED between the two encodes is a site
    /// that missed. That undercounts a site that misses and lands back on the
    /// same width, so it is a lower bound and the header says so.
    private static void PerSite(string id, Dictionary<string, PayloadRow> rows)
    {
        var a = ArmTable.All().First(x => x.Id == id);
        var row = rows[id];
        a.Build();
        var e = Enc.New(Codec.Sites, Math.Max(row.Bytes + 4096, 8192));
        a.ManagedWrite(ref e);
        var after1 = (byte[])e.Widths.Clone();
        e.Reset();
        a.ManagedWrite(ref e);
        long warm = e.PrefixMoves;

        var changed = new List<string>();
#if AK_COUNT
        Console.WriteLine("  {0} warm misses in total, tallied per site by the counting build:", warm);
        for (int i = 0; i < e.SiteMoves.Length; i++)
            if (e.SiteMoves[i] != 0)
                changed.Add(string.Format("    site {0,3}  {1,-46}  {2} miss(es), width now {3}",
                    i, Codec.SiteNames[i], e.SiteMoves[i], e.Widths[i]));
#else
        Console.WriteLine("  {0} warm misses in total. NOT a counting build, so the per-site tally is", warm);
        Console.WriteLine("  unavailable and what follows is the WIDTH DIFF, which undercounts: a site");
        Console.WriteLine("  that misses and lands back on the same width does not appear. Build with");
        Console.WriteLine("  /p:AkCount=true for the real tally.");
        for (int i = 0; i < e.Widths.Length; i++)
            if (e.Widths[i] != after1[i])
                changed.Add(string.Format("    site {0,3}  {1,-46}  width {2} -> {3}",
                    i, Codec.SiteNames[i], after1[i], e.Widths[i]));
#endif
        if (changed.Count == 0) Console.WriteLine("    none");
        else foreach (var c in changed) Console.WriteLine(c);
    }
}
